src/Native/System.Native/pal_networking.cpp

```diff
@@ -428,7 +429,96 @@ static void ConvertHostEntPlatformToPal(HostEntry& hostEntry, hostent& entry)
     }
 }
 
-#if HAVE_GETHOSTBYNAME_R
+#if !HAVE_THREAD_SAFE_GETHOSTBYNAME_AND_GETHOSTBYADDR
+#if !HAVE_GETHOSTBYNAME_R
+static int copy_hostent(struct hostent* from, struct hostent* to,
+                        char* buffer, size_t buflen)
+{
+    // FIXME: the implementation done for this function in https://github.com/dotnet/corefx/commit/6a99b74
+    //        requires testing when managed assemblies are built and tested on NetBSD. Until that time,
+    //        return an error code.
+    return -1;
+}
+
+/*
+Note: we're assuming that all access to these functions are going through these shims on the platforms, which do not provide
+      thread-safe functions to get host name or address. If that is not the case (which is very likely) race condition is
+      possible, for instance; if other libs (such as libcurl) call gethostby[name/addr] simultaneously.
+*/
+static pthread_mutex_t lock_hostbyx_mutex = PTHREAD_MUTEX_INITIALIZER;
```

Comment:
Having a complex-type'd static variable that lives in the STL is very dangerous; there is no deterministic library loading or unloading so there will be a race condition between the process tear down and if the STL is loaded when this variable is being destructed; this could cause crashes on process exit. Please avoid static variables